com.unity.sg2/Editor/GraphUI/EditorCommon/Preview/PreviewManager.cs

```diff
@@ -101,7 +99,142 @@ internal void Initialize(ShaderGraphModel graphModel, MainPreviewView mainPrevie
 
         static bool IsMainContextNode(IGraphElementModel nodeModel)
         {
-            return nodeModel is GraphDataContextNodeModel contextNode && contextNode.graphDataName == new Defs.ShaderGraphContext().GetRegistryKey().Name;
+            return nodeModel is GraphDataContextNodeModel contextNode
+                    && contextNode.graphDataName == new Defs.ShaderGraphContext().GetRegistryKey().Name;
+        }
+
+        internal void UpdateReferencesAfterUndoRedo(
+            GraphModelStateComponent graphModelStateComponent,
+            ShaderGraphModel graphModel)
+        {
+            m_GraphModelStateComponent = graphModelStateComponent;
+            m_GraphModel = graphModel;
+
+            m_PreviewHandlerInstance.SetActiveGraph(m_GraphModel.GraphHandler);
+        }
+
+        /// <summary>
+        ///  Called to reinitialize the textures for the previews when the graph is saved
+        /// </summary>
+        internal void HandleGraphReload(
+            GraphModelStateComponent graphModelStateComponent,
+            ShaderGraphModel graphModel,
+            MainPreviewView mainPreviewView)
+        {
+            m_GraphModelStateComponent = graphModelStateComponent;
+            m_GraphModel = graphModel;
+            m_MainPreviewView = mainPreviewView;
+
+            ReassignAllNodePreviewTextures();
```

Comment:
This function and the one above seem a little weird to me-- specialized handling for cases like these means there's something else about how we have things setup that needs to change.